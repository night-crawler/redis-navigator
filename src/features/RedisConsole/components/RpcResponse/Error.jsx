import { isArray, isPlainObject } from 'lodash';
import PropTypes from 'prop-types';
import React from 'react';
import ReactJson from 'react-json-view';
import { Header } from 'semantic-ui-react';

RpcError.propTypes = {
    error: PropTypes.shape({
        message: PropTypes.string,
        data: PropTypes.any,
        code: PropTypes.any,
    }),
};

export default function RpcError(props) {
    const { error } = props;
    const errorRepr = isArray(error.data) || isPlainObject(error.data)
        ? <ReactJson src={ error.data } name={ false } />
        : <div className='plain-error'>{ error.data }</div>;

    return (
        <div>
            <Header as='h4'>
                <Header.Content>{ error.message }</Header.Content>
                <Header.Subheader>Code: { error.code }</Header.Subheader>
            </Header>
            { errorRepr }
        </div>
    );
}