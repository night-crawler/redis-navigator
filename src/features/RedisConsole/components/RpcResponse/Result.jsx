import { isArray, isBoolean, isPlainObject, isString } from 'lodash';
import PropTypes from 'prop-types';
import React from 'react';
import { Card, Header, Segment } from 'semantic-ui-react';

import { TextareaSpoiler } from '~/features/Common/components';

import { isJson, MimeDetector, isBase64, isValidNumber, isYaml } from '~/utils';

import { BooleanCard } from './BooleanCard';
import { ImageCard } from './ImageCard';
import { ObjectCard } from './ObjectCard';
import { ObjectTreeViewWidget } from './ObjectTreeViewWidget';


StringCard.propTypes = {
  result: PropTypes.string.isRequired,
};
export function StringCard(props) {
  const { result } = props;

  let innerResult = false, type = '';

  if (isBase64(result, 4)) {
    innerResult = atob(result);
    type = 'base64';
  } else if (isValidNumber(result)) {
    // just deny json conversion
  }

  if (innerResult) {
    return (
      <Segment basic={ true }>
        <Header as='h5'>String[{ result.length }], { type }</Header>
        <TextareaSpoiler className='left floated' result={ result } />
        { <Result result={ innerResult } /> }
      </Segment>
    );
  }

  return (
    <Card fluid={ true }>
      <Card.Content>
        <Card.Header>String[{ result.length }]{ type ? ', ' : '' }{ type }</Card.Header>
        <Card.Description>
          <pre>{ result }</pre>
        </Card.Description>
      </Card.Content>
    </Card>
  );
}


Result.propTypes = {
  result: PropTypes.any,
};

export function Result(props) {
  const
    { result } = props,
    detector = new MimeDetector(result);

  if (detector.isImage)
    return <ImageCard dataUri={ detector.imageDataURI } />;

  if (isArray(result))
    return <ObjectTreeViewWidget result={ result } />;

  if (isPlainObject(result))
    return <ObjectTreeViewWidget result={ result } />;

  if (isBoolean(result))
    return <BooleanCard result={ result } />;

  if (isJson(result))
    return <ObjectCard result={ result } mode='json' />;

  if (isYaml(result))
    return <ObjectCard result={ result } mode='yaml' />;

  if (isString(result))
    return <StringCard result={ result } />;

  if (result === null)
    return <div>null</div>;

  return <div>{ result }</div>;
}

