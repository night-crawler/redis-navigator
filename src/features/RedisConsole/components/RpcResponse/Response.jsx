import PropTypes from 'prop-types';
import React from 'react';

import RpcError from './Error';
import Result from './Result';


Response.propTypes = {
  response: PropTypes.shape({
    error: PropTypes.any,
    result: PropTypes.any,
  })
};
export default function Response(props) {
  const { response: { result, error } } = props;
  if (result !== undefined) {
    return <Result result={ result } />;
  }
  if (error !== undefined) {
    return <RpcError error={ error } />;
  }
  return <div>Command has not been executed yet...</div>;
}
