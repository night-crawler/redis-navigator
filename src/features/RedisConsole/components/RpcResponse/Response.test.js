/* eslint-disable react/react-in-jsx-scope,no-undef */
import Response from './Response';
import Result from './Result';
import Error from './Error';


describe('Response', () => {
    it('should render Response.Result', () => {
        const wrapper = shallow(
            <Response
                response={ {
                    result: 1
                } }
            />
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.containsMatchingElement(<Result />)).toEqual(true);
    });


    it('should render Response.Error', () => {
        const wrapper = shallow(
            <Response
                response={ {
                    error: {
                        message: 'test',
                        data: []
                    }
                } }
            />
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.containsMatchingElement(<Error />)).toEqual(true);
    });

});
