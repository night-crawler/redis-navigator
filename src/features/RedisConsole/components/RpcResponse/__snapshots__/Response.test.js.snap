// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response should render Response.Error 1`] = `
<RpcError
  error={
    Object {
      "data": Array [],
      "message": "test",
    }
  }
/>
`;

exports[`Response should render Response.Result 1`] = `
<Result
  result={1}
/>
`;
